### PR TITLE
Add support for manufacturer-specific clusters, commands, and attributes

### LINKED ISSUE #429: Permit manufacturer-specific clusters on an endpoint
I want to work with a device that has a manufacturer-specific cluster (i.e., with a cluster ID in the range reserved for manufacturer-specific clusters, `0xfc00 – 0xffff`).

With the current state of this library, I could work with this cluster by 
* subclassing from `ZclCluster`, and sending commands / accessing attributes via my subclass, and
* registering listeners with the `ZigBeeNetworkManager` to get commands and reports from the cluster.

This is already nice :)

However, I want to use the manufacturer-specific cluster within the openHAB ZigBee binding; the code in this binding relies at some points on accessing clusters via the `ZigBeeEndpoint` of a `ZigBeeNode` - and with the approach described above the cluster will not be available via the `ZigBeeEndpoint`. A concrete example for this:
* The  `ZigBeeDeviceConfigHandler` of the binding accesses attributes via `ZigBeeNode#getEndpoint#get{In|Out}putCluster`. This class, hence, cannot write attributes of the manufacturer-specific cluster, because this cluster is unknown to the `ZigBeeEndpoint`.

One solution to improve here would be to provide the possibility to add clusters to a `ZigBeeEndpoint` (currently, all clusters of a `ZigBeeEndpoint` are derived from the endpoint's simple descriptor, ignoring manufacturer-specific clusters). This method could look as follows:
```java
  /**
     * Add a cluster to this endpoint. This method is intended to be used for clusters that are (a) not reported
     * within the simple descriptor of the endpoint, or (b) not supported directly by the library, and, hence, are not
     * contained in {@link ZclClusterType}.
     */
    public void addCluster(ZclCluster cluster) {
        if (cluster.isClient()) {
            outputClusters.put(cluster.getClusterId(), cluster);
        } else {
            inputClusters.put(cluster.getClusterId(), cluster);
        }
    }
```

Then, in the openHAB binding, one could add manufacturer-specific clusters to an endpoint of a node, e.g., using a custom thing handler.

What do you think, @cdjackson?

### PR DESCRIPTION
**Pull request by triller-telekom**
Replacement for PR #498

Closes #429

Also-by: Henning Sudbrock <henning.sudbrock@telekom.de>
Signed-off-by: Stefan Triller <stefan.triller@telekom.de>

### CODE REVIEW

**Comment by triller-telekom** [comment]
@cdjackson FYI: Here is the commit which adjusts Hennings changes to the current ZSS-1.2.0 version. I have already tested it with a Centralite 3310-D device which provides humidity and temperature values. Please have a look.

**Comment by cdjackson** [comment]
This pull request **introduces 2 alerts** when merging c202c6de1709330bf6c06f02514068ca61d1bcc5 into 339eef60cbc6b3b959ef985685bd61ce23542da8 - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-d95b7dbdf315e09c3d39ad0d91bc30da6d7788fb)

**new alerts:**

* 2 for Spurious Javadoc @param tags

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by cdjackson** [comment]
Thanks @triller-telekom . I had a quick look and it looks similar to what I also did in this branch (https://github.com/zsmartsystems/com.zsmartsystems.zigbee/commit/5f223190a6bed78250f4c8980edea8a5f1598fa4) as I mentioned in my email to Michael last week ;)

The main difference I think is that I don't like the new version of sendDefaultResponse. Given that most commands are not manufacturer specific, I don't think we should require all uses of this to include ```null``` on the end of the call. Maybe you can take a look at this? This should avoid changes to the OTA cluster as well.

I think if we can align on that, then we can merge.

**Comment by triller-telekom** [comment]
I have updated the PR to be rebased on your newly force-pushed dev-1.2.0 branch and I added one commit which reverts the changes to `sendDefaultResponse`.

Basically I extended the existing `sendDefaultResponse` method with the manufacturer specific parameter and made the original method without that parameter call the new one with `null`.

@cdjackson Please have a look at this PR again.

**Comment by cdjackson** [comment]
Sorry @triller-telekom can you rebase this - it's showing too many changes now as I rebased master this morning. If you can rebase on Monday, I'll review without any further changes so we can get this merged.

**Comment by triller-telekom** [comment]
@cdjackson I have rebased the commits again. Please have a look.

**Comment by cdjackson** [comment]
This pull request **introduces 1 alert** when merging 7972c25c773bc8e5b9b714470e23e38708d8ae07 into f054cf9ea0f924e47768d21dc322b01d0ed1f880 - [view on LGTM.com](https://lgtm.com/projects/g/zsmartsystems/com.zsmartsystems.zigbee/rev/pr-dfdc61604c367a15c06637232fa028ca011dcd60)

**new alerts:**

* 1 for Spurious Javadoc @param tags

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by triller-telekom** [comment]
Thank you for your review. I have added a commit to address your comments. Feels free to squash the commits before merging. I just left them separate to simplify your reviewing efforts.

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zcl/ZclCluster.java [hunk 1]
```diff
@@ -1285,4 +1413,50 @@ protected Object readSync(final ZclAttribute attribute) {
...
+     * Returns the manufacturer code; must be non-null for manufacturer-specific clusters.
+     */
+    public Integer getManufacturerCode() {
+        return null;
+    }
+
+    /**
+     * Adds additional attributes to the cluster (like, e.g., manufacturer-specific attributes).
+     */
+    public void addAttributes(Set<ZclAttribute> attributes) {
```
[the remark this answers is not in the document]
Comment (cdjackson) [reply com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zcl/ZclCluster.java]: I think you missed this one?

File: com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zcl/ZclCluster.java [hunk 2]
```diff
@@ -1285,4 +1413,50 @@ protected Object readSync(final ZclAttribute attribute) {
...
+                this.clientAttributes.put(attribute.getId(), attribute);
+            } else {
+                this.serverAttributes.put(attribute.getId(), attribute);
+            }
+        }
+    }
+
+    /**
+     * Adds additional client commands to the cluster (like, e.g., manufacturer-specific commands).
+     */
```
[the remark this answers is not in the document]
Comment (cdjackson) [reply com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/zcl/ZclCluster.java]: I think you missed this one?

**Comment by cdjackson** [comment]
Sorry @triller-telekom github was showing me outdated review data - having sent the comment, I noticed it was kindly telling me it was outdated, where for the others it showed the updated info. Strange...

Anyway - looks good - thanks.

**Comment by triller-telekom** [comment]
Thanks for merging. No worries regarding github. I guess its because your comment was on the method declaration whereas my necessary changes where above that in the comments so in a different line. That's probably why github was not able to match them.

**Pull request merged**